### ISSUE TITLE
No TLS logging and no login failure when TLS is disabled Client side

### ISSUE THREAD

**Issue by gertjanschouten** [neutrinolabs/xrdp#661]
Hello,

I am running XRDP with XFCE and I am trying to secure my connection with TLS.

My xrdp.ini has this:

crypt_level=high
certificate=/etc/xrdp/xrdp.crt
key_file=/etc/xrdp/xrdp.key
security_layer=tls
disableSSLv3=true
tls_ciphers='DHE-RSA-AES128-SHA256 DHE-RSA-AES256-SHA256 DHE-RSA-AES128-GCM-SHA256 DHE-RSA-AES256-GCM-SHA384 ECDHE-RSA-AES128-SHA256 ECDHE-RSA-AES256-SHA384 ECDHE-RSA-AES128-GCM-SHA256 ECDHE-RSA-AES256-GCM-SHA384 ECDHE-ECDSA-AES128-SHA256 ECDHE-ECDSA-AES256-SHA384 ECDHE-ECDSA-AES128-GCM-SHA256 ECDHE-ECDSA-AES256-GCM-SHA384'
LogFile=/etc/xrdp/xrdp.log
LogLevel=DEBUG

My sesman.ini has this:

AllowRootLogin=0
LogFile=/etc/xrdp/sesman.log
LogLevel=DEBUG

However, when I login with Remmina, it _always succeeds_, even if I select RDP security or no security at all. If I try it with the "xfreerdp --authonly --no-tls" command, it also succeeds. I expect login to fail if the client does not want to use TLS.

Moreover, even when the client uses TLS, there is _no logging whatsoever_ on either the client or the server that tells me if and what security is used. For all I know, my connection is completely unsecured.

Does anyone know of any way to _test and verify_ my TLS connection? Also, could functionality be added that extensively logs the details of the TLS authentication process (handshake, protocol and cipher suite selection, etc) at least at DEBUG level and finally, could you please make login **fail** when the client does not want to use TLS?

The AllowRootLogin=0 works, BTW. It's the only thing that I can be sure of that works...

Thanks for your amazing work!

**Comment by gertjanschouten** [comment]
Also, my key and certificate are generated as follows:

openssl req -newkey rsa:2048 -sha512 -x509 -days 99999 -nodes -out xrdp.crt -keyout xrdp.key

And yes, I know that the ECDSA cipher suites aren't going to be used with an RSA key, but that list is the approved list of all cipher suites in our organization and as far as I know, openssl does not mind if you list suites that can't be used with a certain key :-)

**Comment by proski** [comment]
The length of the `tls_ciphers` field in the `struct xrdp_client_info` is limited to 64. The string from the configuration file is copied to that field without any length checking (function `xrdp_rdp_read_config` in `libxrdp/xrdp_rdp.c`). It looks like you are getting a buffer overrun. Also, I don't see any processing for quotes in the code.

**Comment by gertjanschouten** [comment]
Okay, thanks! I will try to keep it within 64. Could this be fixed though? (the length and the quotes) It should be possible to explicitly specify cipher suites instead of using openssl expression language.

Also, when there is a buffer overrun or another error, shouldn't that cause xrdp to either not start or at least log an error instead of silently ignoring all security settings and letting people in unsecured?

I'll change my config to less than 64 and try again if I can login without TLS. I hope not :) Also, I will check if I get some logging this time.

**Comment by gertjanschouten** [comment]
Oh, never mind the quotes, I see that colons are also acceptable. Still, values longer than 64 are still desired, to be able to explicitly select suites.

**Comment by gertjanschouten** [comment]
Okay, if I change it to

tls_ciphers=DHE-RSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256

I get: "Error: protocol security negotiation failure" at the client. In fact, when I comment out every single TLS related line, resulting in:

crypt_level=high
# certificate=/etc/xrdp/xrdp.crt
# key_file=/etc/xrdp/xrdp.key
# security_layer=tls
# disableSSLv3=true
# tls_ciphers=DHE-RSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256
LogFile=/etc/xrdp/xrdp.log
LogLevel=DEBUG

I still get that error. As usual, logs are blank. It seems the only way I can connect is by causing a buffer overrun....  :-S

**Comment by metalefty** [comment]
I also think the maximum length of tls_ciphers should be longer. It 's already filed as #544.

**Comment by metalefty** [comment]
Works for me. The server is configured TLS only. 
```
% xfreerdp -sec-tls /v:server.example.jp
[10:19:52:023] [87564:08816000] [INFO][com.freerdp.utils] - Registering signal hook...
[10:19:52:027] [87564:08816500] [INFO][com.freerdp.client.x11] - No user name set. - Using login name: meta
[10:19:53:604] [87564:08816500] [ERROR][com.freerdp.core] - freerdp_set_last_error ERRCONNECT_TLS_CONNECT_FAILED [0x20008]
[10:19:53:604] [87564:08816500] [ERROR][com.freerdp.core.connection] - Error: protocol security negotiation or connection failure
[10:19:53:604] [87564:08816500] [ERROR][com.freerdp.client.x11] - Freerdp connect error exit status 1
```

**Comment by gertjanschouten** [comment]
Okay, my earlier message about the connection failure was resolved when I reboot the server. Apparently, a simple "service xrdp restart" is not enough to make configuration changes, which can further be demonstrated by doing a "service xrdp stop" followed by a "ps -Af | grep xrdp" which will show that the exact same xrdp processes are still running...

Anyway, a shorter TLS cipher string does not change anything for me. I can still connect without TLS. metalefty, what kind of xfreerdp do you use? My one does not recognize the -sec-tls option, I have to put --no-tls, so it seems that my one (version 1.0.2 on Linux Mint) has a different syntax.

My xrdp.ini is:

[globals]
bitmap_cache=yes
bitmap_compression=yes
port=3389
crypt_level=high
channel_code=1
max_bpp=24

certificate=/etc/xrdp/xrdp.crt
key_file=/etc/xrdp/xrdp.key
security_layer=tls
disableSSLv3=true
tls_ciphers=DHE-RSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256

LogFile=/etc/xrdp/xrdp.log
LogLevel=DEBUG

[xrdp1]
name=PentahoRDP
lib=libvnc.so
username=ask
password=ask
ip=127.0.0.1
port=-1

So, the list of requests is now:
- On the server side, please log the entire SSL handshake procedure, including cipher suite selection
- Also log it if there is no security or RDP security
- Make it possible to have a _testable and verifyable_ TLS connection
- Always fail to login when client does not want to use security protocol that has been made mandatory by server (metalefty has this working with his version of xfreerdp, but it should always work)
- In case of configuration errors _fail to start with an error message_ instead of silently ignoring security
- In case of restart, actually restart, meaning ending all processes and restarting them
- If the server state becomes corrupted, which was probably responsible for my connection failures, please _log the details_ on the server
- Make the cipher string way longer

**Comment by gertjanschouten** [comment]
Metalefty, what does your xrdp.ini look like?

**Comment by metalefty** [comment]
It looks FreeRDP ignoring `--no-tls` option for me, no evidence though.

For command line options, see here: https://github.com/FreeRDP/FreeRDP/wiki/CommandLineInterface

`--no-tls` is an deprecated option to disabling TLS.  I don't correctly remember when the CLI options were changed, but newer version of FreeRDP accepts `-sec-tls`.  Your FreeRDP looks too old.  `+sec-tls` enables TLS, `-sec-tls` disables TLS.

>* On the server side, please log the entire SSL handshake procedure, including cipher suite selection
>* Also log it if there is no security or RDP security

These are already implemented in #426 (after v0.9.1 release though). You'll get the log like this.
```
[20170210-17:05:04] [INFO ] Non-TLS connection established from 192.168.69.1 port 55423 encrypted with standard RDP security
[20170210-17:06:34] [INFO ] TLS connection established from 192.168.69.1 port 52767: TLSv1.2 with cipher AES256-GCM-SHA384
```

> * In case of configuration errors fail to start with an error message instead of silently ignoring security

xrdp never ignore security even if `tls_ciphers` is configured wrong.
For example, `tls_ciphers=WRONG-STATEMENT-HERE`, connection should fail.  FreeRDP outputs following logs.

```
[19:30:18:787] [24777:08816000] [INFO][com.freerdp.utils] - Registering signal hook...
[19:30:18:796] [24777:08816500] [INFO][com.freerdp.client.x11] - No user name set. - Using login name: meta
[19:30:18:141] [24777:08816500] [ERROR][com.freerdp.core] - freerdp_set_last_error ERRCONNECT_TLS_CONNECT_FAILED [0x20008]
[19:30:18:145] [24777:08816500] [ERROR][com.freerdp.core.connection] - Error: protocol security negotiation or connection failure
[19:30:18:146] [24777:08816500] [ERROR][com.freerdp.client.x11] - Freerdp connect error exit status 1
```

My xrdp.ini is like this:
```
security_layer=tls
disableSSLv3=yes
tls_ciphers=HIGH
```

HIGH is equivalent to the result of `openssl ciphers HIGH` command on the environment. In my environment, `HIGH` is equivalent to `ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:SRP-DSS-AES-256-CBC-SHA:SRP-RSA-AES-256-CBC-SHA:SRP-AES-256-CBC-SHA:DH-DSS-AES256-GCM-SHA384:DHE-DSS-AES256-GCM-SHA384:DH-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA256:DH-RSA-AES256-SHA256:DH-DSS-AES256-SHA256:DHE-RSA-AES256-SHA:DHE-DSS-AES256-SHA:DH-RSA-AES256-SHA:DH-DSS-AES256-SHA:DHE-RSA-CAMELLIA256-SHA:DHE-DSS-CAMELLIA256-SHA:DH-RSA-CAMELLIA256-SHA:DH-DSS-CAMELLIA256-SHA:AECDH-AES256-SHA:ADH-AES256-GCM-SHA384:ADH-AES256-SHA256:ADH-AES256-SHA:ADH-CAMELLIA256-SHA:ECDH-RSA-AES256-GCM-SHA384:ECDH-ECDSA-AES256-GCM-SHA384:ECDH-RSA-AES256-SHA384:ECDH-ECDSA-AES256-SHA384:ECDH-RSA-AES256-SHA:ECDH-ECDSA-AES256-SHA:AES256-GCM-SHA384:AES256-SHA256:AES256-SHA:CAMELLIA256-SHA:PSK-AES256-CBC-SHA:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:SRP-DSS-AES-128-CBC-SHA:SRP-RSA-AES-128-CBC-SHA:SRP-AES-128-CBC-SHA:DH-DSS-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:DH-RSA-AES128-GCM-SHA256:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES128-SHA256:DHE-DSS-AES128-SHA256:DH-RSA-AES128-SHA256:DH-DSS-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA:DH-RSA-AES128-SHA:DH-DSS-AES128-SHA:DHE-RSA-CAMELLIA128-SHA:DHE-DSS-CAMELLIA128-SHA:DH-RSA-CAMELLIA128-SHA:DH-DSS-CAMELLIA128-SHA:AECDH-AES128-SHA:ADH-AES128-GCM-SHA256:ADH-AES128-SHA256:ADH-AES128-SHA:ADH-CAMELLIA128-SHA:ECDH-RSA-AES128-GCM-SHA256:ECDH-ECDSA-AES128-GCM-SHA256:ECDH-RSA-AES128-SHA256:ECDH-ECDSA-AES128-SHA256:ECDH-RSA-AES128-SHA:ECDH-ECDSA-AES128-SHA:AES128-GCM-SHA256:AES128-SHA256:AES128-SHA:CAMELLIA128-SHA:PSK-AES128-CBC-SHA`

**Comment by gertjanschouten** [comment]
> xrdp never ignore security even if tls_ciphers is configured wrong
According to proski above, the buffer overrun does not lead to an error

> These are already implemented in #426 (after v0.9.1 release though). You'll get the log like this.
Good to know, thanks! I'll definitely check this release out.

> Your FreeRDP looks too old.
Firstly, what version do you have (xfreerdp --version)? And secondly, security should never depend on the client. Both my old FreeRDP as well as my Remmina can connect without TLS, which shouldn't be possible.

Thanks for all your feedback, it is greatly appreciated :-)

**Comment by metalefty** [comment]
Please try FreeRDP 1.2.0 or later.

> According to proski above, the buffer overrun does not lead to an error

Not leading to an error DOESN'T mean ignoring security. Do you understand that?

>  And secondly, security should never depend on the client. Both my old FreeRDP as well as my Remmina can connect without TLS, which shouldn't be possible.  

I have doubt on this.  The connection is really without TLS? xrdp never accept non-TLS connection if the server is configured TLS only.  Let's check it as the connection type is logged now.

Did you read this? 
> It looks FreeRDP ignoring --no-tls option for me, no evidence though.

So, if you succeed to connect TLS only xrdp server when you specified `--no-tls` option, the connection may be TLS. Because xrdp server never accept non-TLS.  Still no evidence but let's check it.

**Comment by metalefty** [comment]
I confirmed that your version of FreeRDP ignores `--no-tls` option.

Client: CentOS 7, freerdp-1.0.2-10.el7.x86_64
Server: xrdp 8a1de8dbc4310c00ce8fbe939346b55c4db1c2e9 (configured TLS only)

```
client$ xfreerdp --no-tls xrdp-server.example.com
connected to xrdp-server.example.com
creating directory /home/centos/.freerdp/certs
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@           WARNING: CERTIFICATE NAME MISMATCH!           @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
The hostname used for this connection (localhost) 
does not match the name given in the certificate:
localhost.localdomain
A valid certificate for the wrong name should NOT be trusted!
Certificate details:
        Subject: C = AU, ST = Some-State, O = Internet Widgits Pty Ltd, CN = localhost.localdomain
        Issuer: C = AU, ST = Some-State, O = Internet Widgits Pty Ltd, CN = localhost.localdomain
        Thumbprint: 1b:69:aa:12:f7:1e:83:a2:6c:9a:e1:06:d0:a2:48:6b:5c:09:99
The above X.509 certificate could not be verified, possibly because you do not have the CA certificate in your certificate store, or the certificate has expired. Please look at the documentation on how to create local certificate store for a private CA.
Do you trust the above certificate? (Y/N) y
unknown capability type 6
incorrect offset, type:0x06 actual:4 expected:5
```

Certificate warning is shown even though `--no-tls` option is given.  This shoudn't be shown if the connection is actually non-TLS.

On the server side, the log says the connection is **TLSv1 with cipher AES256-SHA**.
```
[20170213-10:07:21] [INFO ] Using default X.509 certificate: /usr/local/etc/xrdp/cert.pem
[20170213-10:07:21] [INFO ] Using default X.509 key file: /usr/local/etc/xrdp/key.pem
[20170213-10:07:21] [DEBUG] Security layer: requested 3, selected 1
[20170213-10:07:30] [INFO ] TLS connection established from ::ffff:127.0.0.1 port 62617: TLSv1 with cipher AES256-SHA
[20170213-10:07:30] [DEBUG] xrdp_0000ed84_wm_login_mode_event_00000001
[20170213-10:07:30] [INFO ] Loading keymap file /usr/local/etc/xrdp/km-00000409.ini
[20170213-10:07:30] [WARN ] local keymap file for 0x00000409 found and doesn't match built in keymap, using local keymap file
```

In conclusion, it's a FreeRDP bug. xrdp never accept non-TLS connections if configured TLS only. The newer versions which has new style command line options can handle `-sec-tls` (formerly `--no-tls`) correctly. You should update FreeRDP.

**Comment by gertjanschouten** [comment]
Okay, will try, thanks.

**Comment by metalefty** [comment]
To conclude, it's a FreeRDP bug. Closing.
